File: code/futureforge/group-34/queue.py
#CEO's Work 20/CSC/119

# A queue is a linear data structure that follows the First In, First Out (FIFO) principle, 
# where elements are inserted at the rear (enqueue) and removed from the front (dequeue) of the queue. 
# It supports operations like enqueue (to insert an element into the queue) and dequeue (to remove and return the front element from the queue).


from collections import deque

class Queue:
    def __init__(self):
        self.items = deque()

    def enqueue(self, item):
        """Inserts an element at the rear of the queue."""
        self.items.append(item)

    def dequeue(self):
        """Removes and returns the front element from the queue."""
        if not self.is_empty():
            return self.items.popleft()
        else:
            raise IndexError("dequeue from empty queue")

    def peek(self):
        """Returns the front element of the queue without removing it."""
        if not self.is_empty():
            return self.items[0]
        else:
            return None

    def is_empty(self):
        """Checks if the queue is empty."""
        return len(self.items) == 0

    def size(self):
        """Returns the number of elements in the queue."""
        return len(self.items)

# Example usage of Queue
queue = Queue()

queue.enqueue("apple")
queue.enqueue("banana")
queue.enqueue("cherry")

print("Queue size:", queue.size())  # Output: 3
print("Front element:", queue.peek())  # Output: apple

print(queue.dequeue())  # Output: apple
print(queue.dequeue())  # Output: banana
print(queue.is_empty())  # Output: False
print(queue.dequeue())  # Output: cherry
print(queue.is_empty())  # Output: True

File: code/futureforge/group-31/queue.js
//Section worked on by Moses Amarachi 20/CSC/078

// Queue
// A queue is a linear data structure that follows the First In, First Out (FIFO) principle, 
// where elements are added to the rear (enqueue) and removed from the front (dequeue) of the queue.

class Queue {
    constructor() {
        this.items = []; // Initialize an empty array to store queue elements
    }
  
    // Add element to the rear of the queue
    enqueue(element) {
        this.items.push(element);
    }
  
    // Remove and return the front element of the queue
    dequeue() {
        if (this.isEmpty()) {
            return "Underflow - Queue is empty";
        }
        return this.items.shift();
    }
  
    // Return the front element of the queue without removing it
    front() {
        if (this.isEmpty()) {
            return "Queue is empty";
        }
        return this.items[0];
    }
  
    // Check if the queue is empty
    isEmpty() {
        return this.items.length === 0;
    }
  
    // Return the size of the queue
    size() {
        return this.items.length;
    }
  
    // Print all elements of the queue
    printQueue() {
        let result = "";
        for (let i = 0; i < this.items.length; i++) {
            result += this.items[i] + " ";
        }
        console.log("Queue: " + result);
    }
  }
  
  // Example usage of the Queue class
  let queue = new Queue();
  queue.enqueue(10);
  queue.enqueue(20);
  queue.enqueue(30);
  
  console.log("Queue size:", queue.size()); // Output: 3
  console.log("Front element:", queue.front()); // Output: 10
  queue.printQueue(); // Output: Queue: 10 20 30
  console.log("Dequeued element:", queue.dequeue()); // Output: 10
  queue.printQueue(); // Output: Queue: 20 30
  
  
  
  